### PR TITLE
Multiple commits

### PR DESCRIPTION
**Pull request by peter1000**

### CODE REVIEW

**Comment by peter1000** [comment]
The idea is that in the future other formts (html) can / should use/share the `render.jl`
**Entries** and related functions: **has_items, mainsetup, prepare_entries, generate_html_id**. This will reduce the code for the individual formats and all can share for instance nice permalink names instead of the ugly once I use now for html.

**Comment by peter1000** [comment]
**after this is merged** - we still need to do some documentation adjustments and maybe add again some examples.

**Comment by peter1000** [comment]
**Now I got it!**

I understand now what  you wanted to say by **This will mutate config which probably isn't the behaviour we want?** https://github.com/MichaelHatherly/Lexicon.jl/pull/88#discussion-diff-29103252 

It was first not clear to me - sorry taking that long. 


I finished pushing to this - can be merged. After this is merged - we still need to do some documentation adjustments and maybe add again some examples.

File: src/render.jl [hunk 1]
```diff
@@ -5,72 +5,169 @@ parsedocs(ds::Docile.Interface.Docs{:md}) = Markdown.parse(data(ds))
...
+                basename = "comment"
+            else
+                # adjust basename to be a valid html id
+                basename = generate_html_id(string(k == :macro ? macroname(ent) : name(obj)))
+            end
+            basename in keys(basenames)                  ?
+                    anchornum = basenames[basename] += 1 :
+                    anchornum = basenames[basename] = 1
+            # string(k, "_", generate_html_id(s))
+            string(k, "_", generate_html_id(s))
```
Comment (MichaelHatherly) [inline src/render.jl]: Leftover debugging code?

File: src/render.jl [hunk 2]
```diff
@@ -5,72 +5,169 @@ parsedocs(ds::Docile.Interface.Docs{:md}) = Markdown.parse(data(ds))
...
+                    "$(string(k))__$(basename).$(anchornum)", k)
+        end
+    end
+    return ents
 end
 
 file"docs/save.md"
-function save(file::AbstractString, modulename::Module; args...)
-    config = Config(; args...)
+function save(file::AbstractString, modulename::Module, config::Config; args...)
```
Comment (MichaelHatherly) [inline src/render.jl]: Having an optional argument for `config`, ie. `config::Config = Config()`, here and in the other `save` method below would avoid needing to *always* provide a `Config` object. That would result in slightly cleaner user build scripts if they don't want to change any settings.

**Comment by MichaelHatherly** [comment]
Other than the two comments this is good to go. Presumably tests covering the new features will be added with the docs updates in the next PR?

**Comment by peter1000** [comment]
Will changed this things and a couple of small other leftovers.

Will later continue to push here.

* I will change the `genindex` to just ` index` - from your comments I seem to get the feeling you prefer it that way and it is also shorter e.g. https://github.com/MichaelHatherly/Lexicon.jl/pull/89#discussion-diff-29107624R36

* I will change the type for `md_index_modprefix` https://github.com/MichaelHatherly/Lexicon.jl/pull/89#discussion-diff-29107624R24 to `ByteString`.  

  `UTF8String` gives a warning TypeError and `ASCIIString` is not good enough for some languages. `ByteString` is `Union(ASCIIString,UTF8String)` and should be just fine and works for julia 0.3 and 0.4

**Question about the docs:** We have the long `Config` options with the `save` documentation.
I thought now where `Config` is exported it might be more suitable to associate the configuration options with the type `Config` itself.

**Comment by MichaelHatherly** [comment]
> `UTF8String` gives a warning `TypeError` and `ASCIIString` is not good enough for some languages.

Try this diff:

```diff
diff --git a/src/render.jl b/src/render.jl
index e73e059..77b6815 100644
--- a/src/render.jl
+++ b/src/render.jl
@@ -21,7 +21,7 @@ type Config
     mdstyle_subheader      :: ASCIIString
     mdstyle_genindex_mod   :: ASCIIString
     md_subheader           :: Symbol
-    md_genindex_modprefix  :: ASCIIString
+    md_genindex_modprefix  :: UTF8String
     md_permalink           :: Bool
 
     const defaults = Dict{Symbol, Any}([
@@ -46,10 +46,10 @@ end
 
 function update_config!(config::Config, args::Dict)
     for (k, v) in args
-        try
-            k in fieldnames(Config) ? setfield!(config, k, v) : warn("Invalid setting: '$(k) = $(v)'.")
-        catch err # e.g. TypeError
-            warn("Invalid setting: '$(k) = $(v)'. Error: $err")
+        if k ∈ fieldnames(Config)
+            setfield!(config, k, convert(fieldtype(Config, k), v))
+        else
+            warn("Invalid setting: '$(k) = $(v)'.")
         end
     end
```

The problem was the `try`-block catching *every* type of error. This was hiding the need to explicitly call `convert` to get from `ASCIIString` to `UTF8String`.

**Comment by MichaelHatherly** [comment]
> Question about the docs: We have the long `Config` options with the `save` documentation.
> I thought now where `Config` is exported it might be more suitable to associate the configuration options with the type `Config` itself.

Yes, that seems a better place for them.

**Comment by peter1000** [comment]
Should I remove the mention of the package `Sims.jl` as last time I tried to run the docs it through an error: was still using `@doc*` instead of `@doc+` https://github.com/tshort/Sims.jl/blob/master/src/main.jl#L107

```
A more customized build process can be found in the Sims.jl package.
```

Anyway maybe I add instead a link to Dociles README: https://github.com/MichaelHatherly/Docile.jl#projects-using-docile--lexicon

Any thoughts?

**Comment by MichaelHatherly** [comment]
Sims.jl is still relevant. The `*` -> `+` change is just a syntactic one and doesn't mean that the package isn't a value resource for learning from. That section can probably stay as is for the moment.

(Also, it goes without saying really, but current build scripts should continue to function the same as before. Currently this branch has the following error:

```
julia> save("/home/mike/lexicon.md", Lexicon)
ERROR: MethodError: `save` has no method matching save(::ASCIIString, ::Module)
Closest candidates are:
  save(::AbstractString, ::Module, ::Lexicon.Config)
  save(::AbstractString, ::Lexicon.Index, ::Lexicon.Config)
  save(::AbstractString, ::Base.Multimedia.MIME{symbol("text/html")}, ::Docile.Metadata, ::Lexicon.Config)
```

which can probably be fixed with making the `Config` object optional I believe?)

**Comment by peter1000** [comment]
> Currently this branch has the following error:

I know - added a number of commits already to address such will push later.

**Comment by MichaelHatherly** [comment]
> I know - added a number of commits already to address such will push later.

Great, thanks.

**Comment by peter1000** [comment]
this fails in julia 0.3

```
+        if k ∈ fieldnames(Config)
+            setfield!(config, k, convert(fieldtype(Config, k), v))
+        else
+            warn("Invalid setting: '$(k) = $(v)'.")
```

ERROR: type DataType has no field md_permalink

**Comment by peter1000** [comment]
Pushed commits will probably still fail on julia 0.3 needs also some tests and going through the examples. Better also not to TAG too early we might still find some issues when I add some real examples.

**Comment by MichaelHatherly** [comment]
> this fails in julia 0.3

It would appear that in 0.3 `fieldtype` has signature `fieldtype(::T, ::Symbol)` whereas in 0.4 it has `fieldtype(::Type{T}, ::Symbol)`. Solvable with a conditional definition such as

```julia
if VERSION < v"0.4-"
    ft(dt, s) = fieldtype(dt, s)
else
    ft{T}(dt::T, s) = fieldtype(T, s)
end
```

and then use `ft(config, k)` instead of `fieldtype(config, k)`/`fieldtype(Config, k)`.

My *feeling* is that it's academic as to which we use - `ByteString` or `UTF8String` and `convert`. Though the lack of explicit `convert` may cause strange errors later on... possibly.

For the moment, go with which ever solution seems more satisfying to you. If problems come up later, we'll know where to look in the code.

**Comment by peter1000** [comment]
I change your git command to:

```
git subtree push --prefix site origin gh-pages
```

the orginal does not work here as in the docs

```
 git subtree push --prefix docs/build origin gh-pages
'docs/build' does not exist; use 'git subtree add'
```

**Comment by MichaelHatherly** [comment]
> I change your git command

Good catch, thanks!

**Comment by peter1000** [comment]
sorry ;) I use here **ByteString* *- if you want to change it later no problem .. the suggested solution did not work out. Will just go once again through the couple of examples and than Push for merge.

If you do not TAG we can later fix up anything still left or add more examples and some tests.

**Comment by peter1000** [comment]
this example as is is not runable

```
EachEntry(docs::Metadata; order = [:category, :name, :source])
```

should we leave it as such or add things that one can really run in the RPL? Any suggestions

**Comment by MichaelHatherly** [comment]
That whole docstring could probably be better attached to `function EachEntry` below where it currently is. The first example is just an example of the signature for the public method, so it can stay.

(The general consensus seems to be moving in the direction of having the signature to display for a docstring automatically taken from the code block listed at the top of a docstring if it's available, otherwise get the signature from the object itself. The infrastructure isn't in place yet to do this though.)

**Comment by peter1000** [comment]
> The general consensus seems to be moving in the direction of having the signature to display for a docstring automatically taken from the code block listed at the top of a docstring

would be nice

Will change the docstring you mentioned above.

Without the change here are the latest docs http://peter1000.github.io/Lexicon.jl/

produces by the included example

```
using Lexicon
index = save("docs/api/Lexicon.md", Lexicon);
save("docs/api/index.md", Index([index]); md_subheader = :category);
run(`mkdocs gh-deploy`)

```

**Comment by peter1000** [comment]
If that passes I think we can merge and solve anything else as it comes up: I opened two related issues which we might want to consider .

**Comment by MichaelHatherly** [comment]
> Without the change here are the latest docs http://peter1000.github.io/Lexicon.jl/

Very nice! Great work.

(Only nitpick would be the placement of "[Exported]" and "[Internal]" labels (they're controlled by the `:md_subheader` setting right?), but that's really a stylistic issue and won't hold up merging of this.)

**Comment by MichaelHatherly** [comment]
It's passed on traivs, that's good enough for me. AppVeyor will build master again anyway once this is merged.

**Comment by MichaelHatherly** [comment]
@peter1000, the effort you've put in here is much appreciated. Thanks!

**Comment by peter1000** [comment]
> (Only nitpick would be the placement of "[Exported]" and "[Internal]" labels (they're controlled by the :md_subheader setting right?), but that's really a stylistic issue and won't hold up merging of this.)

* Not sure what you would prefer: the default of  `Config` could always be adjusted (just not too often ;) and otherwise we can overwrite it.

**Anyway - thanks a lot for your supervision - I appreciate it.***

After lunch I will look to add an example folder again. Maybe some other issue will come up.

**Comment by peter1000** [comment]
UNRELATED: I looked through the list of projects which require Docile/Lexicon you once posted.

I asked some if I may add them to [Projects using Docile / Lexicon](https://github.com/MichaelHatherly/Docile.jl#projects-using-docile--lexicon)

But what I really wanted to say: most of them which do produce a webpage use actual your `html` format.  Just for info ;)  Did not want to open an other issue for this.

**Pull request merged**